docs(marker/copy): provide example for `&T` being `Copy`

janriemer opened the pull request:
### Edited 2020-08-16 (most recent)
In the current documentation about the `Copy` marker trait, there is a section
with examples of structs that can implement `Copy`. Currently there is no example for
showing that shared references (`&T`) are also `Copy`.
It is worth to have a dedicated example for `&T` being `Copy`, because shared
references are an integral part of the language and it being `Copy` is not as
intuitive as other types that share this behaviour like `i32` or `bool`.

The example picks up on the previous non-`Copy` struct and shows that
structs can be `Copy`, even when they hold a shared reference to a non-`Copy` type.

-----------------------------------------
### Edited 2020-08-02, 3:28 p.m.
I've just realized that it says "in addition to the **implementors listed below**", which makes this PR kind of "wrong", because `&T` is indeed in the "implementors listed below".
Maybe we can instead show an example with `&T` in the [When can my type be Copy](https://doc.rust-lang.org/std/marker/trait.Copy.html#when-can-my-type-be-copy) section.

What I really want to achieve is that it becomes more obvious that `&T` is also `Copy`, because, I think, it is very valuable to know and it wasn't obvious for me, until I read something about it in a forum post.

What do you think? I would create another PR for that.
**Please feel free to close this PR.**

-----------------------------------
### Original post
In the current documentation about the `Copy` marker trait, there is a section
about "additional implementors", which list additional implementors of the `Copy` trait.
The fact that shared references are also `Copy` is mixed with another point,
which makes it hard to recognize and make it seem not as important.

This clarifies the fact that shared references are also `Copy`, by mentioning it as a
separate item in the list of "additional implementors".

janriemer committed:
docs(marker/copy): clarify that `&T` is also `Copy`

In the current documentation about the `Copy` marker trait, there is a section
about "additional implementors", which list additional implementors of the `Copy` trait.
The fact that shared references are also `Copy` is mixed with another point,
which makes it hard to recognize and make it seem not as important.

This clarifies the fact that shared references are also `Copy`, by mentioning it as a
separate item in the list of "additional implementors".

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @kennytm (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

jyn514 commented:
Feel free to update this PR instead of making a new one, I think it would be nice to keep the discussion in one place.

janriemer committed:
docs(marker/copy): provide example for `&T` being `Copy`

In the current documentation about the `Copy` marker trait, there is a section
with examples of structs that can implement `Copy`. Currently there is no example for
showing that shared references (`&T`) are also `Copy`.
It is worth to have a dedicated example for `&T` being `Copy`, because shared
references are an integral part of the language and it being `Copy` is not as
intuitive as other types that share this behaviour like `i32` or `bool`.

The example picks up on the previous non-`Copy` struct and shows that
structs can be `Copy`, even when they hold a shared reference to a non-`Copy` type.

jyn514 commented on library/core/src/marker.rs:
````diff
@@ -315,6 +315,18 @@ pub trait StructuralEq {
 /// the trait `Copy` may not be implemented for this type; field `points` does not implement `Copy`
 /// ```
 ///
+/// Shared references (`&T`) are also `Copy`, so a struct can be `Copy`, even when it holds
````
```suggestion
/// Shared references (`&T`) are also `Copy`, so a type can be `Copy`, even when it holds
```

jyn514 commented on library/core/src/marker.rs:
```diff
@@ -540,7 +552,7 @@ macro_rules! impls {
 /// [the Nomicon](../../nomicon/phantom-data.html).
 ///
-/// # A ghastly note 👻👻👻
+/// # A ghastly note 
```
Any reason this was removed?
```suggestion
/// # A ghastly note 👻👻👻
```

janriemer replied:
Oh no, why has it removed the cute emojis? Thanks for catching that. I'm really sorry. :cry: 
I can only imagine that it has something to do with my EmojiOne extension for Firefox.

janriemer committed:
rephrase: struct -> type

Co-authored-by: Joshua Nelson <joshua@yottadb.com>

janriemer committed:
add back emojis that have been removed accidentally

Co-authored-by: Joshua Nelson <joshua@yottadb.com>

jyn514 commented on library/core/src/marker.rs:
```diff
@@ -540,7 +552,7 @@ macro_rules! impls {
 /// [the Nomicon](../../nomicon/phantom-data.html).
 ///
-/// # A ghastly note 👻👻👻
+/// # A ghastly note emoji unicode: 1f47bemoji unicode: 1f47bemoji unicode: 1f47b
```
🤦 apparently github suggestions don't work with emojis. Do you mind making the change locally?

janriemer replied:
:laughing: Thanks for catching that and your super fast responses. :kissing_heart:
I'll make the changes locally, now (first time, I am pulling the Rust code base on my computer, hui :smiling_face_with_three_hearts: )

janriemer commented:
So ironic that the *ghost* emoji has disappeared. :laughing: But, it is now back! :smiley:
@jyn514 Thank you for your fast responses. :kissing_heart:

jyn514 commented:
No problem :)

Maybe @poliorcetics is interested in reviewing the doc change?

poliorcetics commented:
I'll have a look later. :)

poliorcetics commented on library/core/src/marker.rs:
```diff
@@ -315,6 +315,18 @@ pub trait StructuralEq {
+/// Shared references (`&T`) are also `Copy`, so a type can be `Copy`, even when it holds
+/// shared references of types `T` that are *not* `Copy`. Consider the following struct,
+/// which can implement `Copy`, because it only holds a *shared reference* to our non-`Copy`
+/// type `PointList` from above:
```
The documentation usually separates code block with an empty line above and below for clarity.

```suggestion
/// type `PointList` from above:
///
```

janriemer replied:
Good catch! Thanks.

poliorcetics commented on library/core/src/marker.rs:
````diff
@@ -315,6 +315,18 @@ pub trait StructuralEq {
+/// ```
+/// # #![allow(dead_code)]
+/// # struct PointList;
+/// struct PointListWrapper<'a> {
````
Adding the `derive` would be nice: since examples are compiled and run, showing it works by implementing it could be nice.

```suggestion
/// #[derive(Copy, Clone)]
/// struct PointListWrapper<'a> {
```

janriemer replied:
That is a good idea, thanks for the suggestion! :+1: 
Do you think we should simultaneously add the same `derive` to the [`Point` struct in line 294](https://github.com/rust-lang/rust/blob/009551f758d1d007ad0f7b652bfa8ddba0738117/library/core/src/marker.rs#L294) for consistency?

janriemer replied:
@poliorcetics with the two latest commits this is now resolved.

janriemer committed:
add empty line above code block

Co-authored-by: Poliorcetics <poliorcetics@users.noreply.github.com>

janriemer committed:
docs: add `derive` for struct

Code blocks in doc comments are compiled and run, so we show `Copy` works in this example.

Co-authored-by: Poliorcetics <poliorcetics@users.noreply.github.com>

poliorcetics commented:
Looks good to me, thanks !

I don't have the right to `r+` for bors though, you'll have to wait for someone who does. 😄

janriemer commented:
@poliorcetics Thank you for your swift review! I really appreciate it! :heart: 
[My comment on the `derive`](https://github.com/rust-lang/rust/pull/75049#discussion_r471154122) might have been hidden from you, because the conversation has been marked as "resolved" (I now have "unresolved" it), so I repeat it here. :wink:
>Do you think we should simultaneously add the same `derive` to the [Point struct in line 294](https://github.com/rust-lang/rust/blob/009551f758d1d007ad0f7b652bfa8ddba0738117/library/core/src/marker.rs#L294) for consistency?

poliorcetics commented:
Oh yeah didn't see it sorry. Adding it above too is a good idea !

Commit:
docs: add another `derive` for `Copy`able struct

This adds another `derive` for a `Copy`able struct, so that we are
consistent with `derive` annotations.

janriemer commented:
@poliorcetics cc @jyn514 Sorry, for the late response. All the necessary changes from your review are now present.

jyn514 commented:
@bors r=poliorcetics rollup

bors commented:
:pushpin: Commit 522d177f34e683b16d3548446aa3ff7ddd2c531c has been approved by `poliorcetics`

<!-- @bors r=poliorcetics 522d177f34e683b16d3548446aa3ff7ddd2c531c -->
<!-- homu: {"type":"Approved","sha":"522d177f34e683b16d3548446aa3ff7ddd2c531c","approver":"poliorcetics"} -->

janriemer commented:
@poliorcetics @jyn514 Thank you for your swift responses and this fast merge!❤ It was my first contribution to Rust and I've felt very welcome. I hope, I can contribute some more in the future.🤗
You are all doing a great job for Rust and the open source community! :+1:

jyn514 commented:
No problem! Thanks for your contribution :)

poliorcetics commented:
Thanks for the contribution :) if you continue to do docs PR you'll probably see us often !

Welcome to the open source world !
